In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from tabulate import tabulate
import warnings
warnings.filterwarnings('ignore')

df = pd.read_excel("DS1_C4_S9_Car_Data (1).xlsx")
df

S. No.        Make      Model        Variant  Displacement  Cylinders  \
0          0        Tata  Nano Genx             Xt         624.0        2.0   
1          1        Tata  Nano Genx             Xe         624.0        2.0   
2          2        Tata  Nano Genx        Emax Xm         624.0        2.0   
3          3        Tata  Nano Genx            Xta         624.0        2.0   
4          4        Tata  Nano Genx             Xm         624.0        2.0   
...      ...         ...        ...            ...           ...        ...   
1271    1271       Honda       City   Vx Mt Diesel        1498.0        4.0   
1272    1272       Honda       City   Zx Mt Diesel        1498.0        4.0   
1273    1273       Honda       City  Zx Cvt Petrol        1497.0        4.0   
1274    1274       Honda       City   V Cvt Petrol        1497.0        4.0   
1275    1275  Mitsubishi    Montero         3.2 At        3200.0        4.0   

      Valves_Per_Cylinder  Fuel_Tank_Capacity_litre Fuel_Type  Height_mm  ...  \
0                     2.0                      24.0    Petrol     1652.0  ...   
1                     2.0                      24.0    Petrol     1652.0  ...   
2                     2.0                      15.0       CNG     1652.0  ...   
3                     2.0                      24.0    Petrol     1652.0  ...   
4                     2.0                      24.0    Petrol     1652.0  ...   
...                   ...                       ...       ...        ...  ...   
1271                  4.0                      40.0    Diesel     1495.0  ...   
1272                  4.0                      40.0    Diesel     1495.0  ...   
1273                  4.0                      40.0    Petrol     1495.0  ...   
1274                  4.0                      40.0    Petrol     1495.0  ...   
1275                  4.0                      88.0    Diesel     1900.0  ...   

      Central_Locking  Child_Safety_Locks Hill_Assist  Drive_Modes  \
0                 Yes                 Yes         NaN          NaN   
1                 NaN                 Yes         NaN          NaN   
2                 NaN                 Yes         Yes          NaN   
3                 Yes                 Yes         NaN          NaN   
4                 NaN                 Yes         NaN          NaN   
...               ...                 ...         ...          ...   
1271              Yes                 Yes         NaN          NaN   
1272              Yes                 Yes         NaN          NaN   
1273              Yes                 Yes         NaN          NaN   
1274              Yes                 Yes         NaN          NaN   
1275              Yes                 Yes         Yes          NaN   

      High_Speed_Alert_System  ABS_(Anti-lock_Braking_System)  \
0                         NaN                             NaN   
1                         NaN                             NaN   
2                         NaN                             NaN   
3                         NaN                             NaN   
4                         NaN                             NaN   
...                       ...                             ...   
1271                      Yes                             Yes   
1272                      Yes                             Yes   
1273                      Yes                             Yes   
1274                      Yes                             Yes   
1275                      NaN                             Yes   

     Gross_Vehicle_Weight                                            Airbags  \
0                     NaN                                                NaN   
1                     NaN                                                NaN   
2                     NaN                                                NaN   
3                     NaN                                                NaN   
4                     NaN                                     

In [94]:
#Task 1:
df = pd.DataFrame(data)

# Map 'Yes' to 1 and 'No' to 0, and fill NaNs with 0
df['ABS_(Anti-lock_Braking_System)'] = df['ABS_(Anti-lock_Braking_System)'].map({'Yes': 1, 'No': 0}).fillna(0)
df['Airbags'] = df['Airbags'].map({'Yes': 1, 'No': 0}).fillna(0)
df['Hill_Assist'] = df['Hill_Assist'].map({'Yes': 1, 'No': 0}).fillna(0)

# Function to assess safety for a given car model
def assess_safety(row):
    abs_present = "ABS_(Anti-lock_Braking_System)" if row['ABS_(Anti-lock_Braking_System)'] == 1 else "No ABS"
    airbags_present = "Airbags" if row['Airbags'] == 1 else "No Airbags"
    hill_assist_present = "Hill_Assist" if row['Hill_Assist'] == 1 else "No Hill Assist"
    
    # Safety score: count how many of the features are present
    safety_score = row['ABS_(Anti-lock_Braking_System)'] + row['Airbags'] + row['Hill_Assist']
    
    # Output the result for this car
    print(f"{row['Make']} {row['Model']} has: {abs_present}, {airbags_present}, {hill_assist_present} | Safety Score: {safety_score}/3")

# Apply the function to each row in the DataFrame
print("Safety Feature Assessment:")
df.apply(assess_safety, axis=1)

Safety Feature Assessment:
Tata Nano Genx has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Tata Nano Genx has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Tata Nano Genx has: No ABS, No Airbags, Hill_Assist | Safety Score: 1.0/3
Tata Nano Genx has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Tata Nano Genx has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Tata Nano Genx has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Datsun Redi-Go has: ABS_(Anti-lock_Braking_System), No Airbags, No Hill Assist | Safety Score: 1.0/3
Datsun Redi-Go has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Datsun Redi-Go has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Datsun Redi-Go has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Datsun Redi-Go has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Datsun Redi-Go has: No ABS, No Airbags, No Hill Assist | Safety Score: 0.0/3
Renault Kwid has: ABS_(Anti-

0       None
1       None
2       None
3       None
4       None
        ... 
1271    None
1272    None
1273    None
1274    None
1275    None
Length: 1276, dtype: object

#### Interpretation: the presence or absencs of the ABS, hiss-assist, and safety score are calculated for the different model

In [25]:
#Task 2:
df = pd.DataFrame(data)

# Task 1: Calculate average seating capacity
average_seating_capacity = df['Seating_Capacity'].mean(skipna=True)
print(f"Average Seating Capacity: {average_seating_capacity}")

df['Central_Locking'].fillna('No', inplace=True)
df['Child_Safety_Locks'].fillna('No', inplace=True)

# Task 2: Count occurrences of central locking
central_locking_count = df['Central_Locking'].value_counts()['Yes']  # sum() works because True is treated as 1 and False as 0
print(f"Number of models with Central Locking: {central_locking_count}")

# Task 3: Count occurrences of child safety locks
child_safety_locks_count = df['Child_Safety_Locks'].value_counts()['Yes']
print(f"Number of models with Child Safety Locks: {child_safety_locks_count}")

# Optional: Display the dataframe for reference
print("\nCar Models and Their Features:")
print(df)

Average Seating Capacity: 5.27007874015748
Number of models with Central Locking: 1127
Number of models with Child Safety Locks: 1201

Car Models and Their Features:
      S. No.        Make      Model        Variant  Displacement  Cylinders  \
0          0        Tata  Nano Genx             Xt         624.0        2.0   
1          1        Tata  Nano Genx             Xe         624.0        2.0   
2          2        Tata  Nano Genx        Emax Xm         624.0        2.0   
3          3        Tata  Nano Genx            Xta         624.0        2.0   
4          4        Tata  Nano Genx             Xm         624.0        2.0   
...      ...         ...        ...            ...           ...        ...   
1271    1271       Honda       City   Vx Mt Diesel        1498.0        4.0   
1272    1272       Honda       City   Zx Mt Diesel        1498.0        4.0   
1273    1273       Honda       City  Zx Cvt Petrol        1497.0        4.0   
1274    1274       Honda       City   V Cvt 

#### Interpretation: Average Seating Capacity: 5.27007874015748 , Number of models with Central Locking: 112 ,  Number of models with Child Safety Locks: 1201

In [63]:
#Task 3:
# Create DataFrame
df = pd.DataFrame(data)

# Task 1: Handle missing values (Fill with 'No' for simplicity, can also use 'Unknown')
df['High_Speed_Alert_System'].fillna('No', inplace=True)
df['Fasten_Seat_Belt_Warning'].fillna('No', inplace=True)
df['Door_Ajar_Warning'].fillna('No', inplace=True)

# Task 2: Count occurrences of each feature
high_speed_alerts_count = df['High_Speed_Alert_System'].value_counts()['Yes']
seat_belt_reminders_count = df['Fasten_Seat_Belt_Warning'].value_counts()['Yes']
door_ajar_warnings_count = df['Door_Ajar_Warning'].value_counts()['Yes']

# Task 3: Calculate percentage of models with each feature
total_models = len(df)
high_speed_alerts_percentage = (high_speed_alerts_count / total_models) * 100
seat_belt_reminders_percentage = (seat_belt_reminders_count / total_models) * 100
door_ajar_warnings_percentage = (door_ajar_warnings_count / total_models) * 100

# Output the results
print(f"Number of models with High-Speed Alerts: {high_speed_alerts_count} ({high_speed_alerts_percentage:.2f}%)")
print(f"Number of models with Seat Belt Reminders: {seat_belt_reminders_count} ({seat_belt_reminders_percentage:.2f}%)")
print(f"Number of models with Door-Ajar Warnings: {door_ajar_warnings_count} ({door_ajar_warnings_percentage:.2f}%)")

# Optional: Display the dataframe for reference
print("\nCar Models and Their Alert Systems Features:")
print(df)

Number of models with High-Speed Alerts: 208 (16.30%)
Number of models with Seat Belt Reminders: 1086 (85.11%)
Number of models with Door-Ajar Warnings: 1133 (88.79%)

Car Models and Their Alert Systems Features:
      S. No.        Make      Model        Variant  Displacement  Cylinders  \
0          0        Tata  Nano Genx             Xt         624.0        2.0   
1          1        Tata  Nano Genx             Xe         624.0        2.0   
2          2        Tata  Nano Genx        Emax Xm         624.0        2.0   
3          3        Tata  Nano Genx            Xta         624.0        2.0   
4          4        Tata  Nano Genx             Xm         624.0        2.0   
...      ...         ...        ...            ...           ...        ...   
1271    1271       Honda       City   Vx Mt Diesel        1498.0        4.0   
1272    1272       Honda       City   Zx Mt Diesel        1498.0        4.0   
1273    1273       Honda       City  Zx Cvt Petrol        1497.0        4.0 

#### Interpretation: Number of models with High-Speed Alerts: 208 (16.30%),   Number of models with Seat Belt Reminders: 1086 (85.11%),   Number of models with Door-Ajar Warnings: 1133 (88.79%)

In [44]:
#Task 4:
df = pd.DataFrame(data)

# Task 1: Analyze overall average dimensions (height, length, width)
average_height = df['Height_mm'].mean()
average_length = df['Length_mm'].mean()
average_width = df['Width_mm'].mean()

print(f"Overall Average Height: {average_height:.2f} mm")
print(f"Overall Average Length: {average_length:.2f} mm")
print(f"Overall Average Width: {average_width:.2f} mm")

# Task 2: Make-wise average summaries for dimensions
make_wise_avg = df.groupby('Make').mean(numeric_only=True).reset_index()

# Rename columns for better readability
make_wise_avg.rename(columns={'Height_mm': 'Avg Height (mm)', 'Length_mm': 'Avg Length (mm)', 'Width_mm': 'Avg Width (mm)'}, inplace=True)

print("\nMake-wise Average Dimensions:")
print(make_wise_avg)

# Optional: Display the original dataframe for reference
print("\nCar Models and Their Dimensions:")
print(df)

Overall Average Height: 1592.54 mm
Overall Average Length: 4296.88 mm
Overall Average Width: 1788.55 mm

Make-wise Average Dimensions:
                Make       S. No.  Displacement  Cylinders  \
0       Aston Martin   820.000000   5038.333333  10.666667   
1               Audi   522.387097   2256.354839   5.161290   
2              Bajaj   357.500000    216.000000        NaN   
3            Bentley   652.166667   5447.833333  10.666667   
4                Bmw   719.702703   2707.918919   5.459459   
5            Bugatti   355.500000   7993.000000  16.000000   
6             Datsun    26.800000   1065.066667   3.000000   
7                 Dc   262.000000   2000.000000   4.000000   
8            Ferrari   802.000000   4589.500000   8.500000   
9               Fiat   668.652174   1269.391304   4.000000   
10             Force   633.500000   2447.000000   4.000000   
11              Ford   429.837209   1551.325581   3.883721   
12             Honda   674.484375   1444.875000   4.000000 

#### Interpretation: Overall Average Height: 1592.54 mm , Overall Average Length: 4296.88 m ,
Overall Average Width: 1788.55 mmmm

### Task 5
#### Overall interpretation for sprint 9 and 10


1) The most prefered car body type is SUV
2) Their are zero number of cars accomadating all the driving modes
3) the most preferd car based on the city milage is Maruti Suzaki(Non elcetric vehical)
4) The mean and medain  for hyundai is approximately similar fro both the city milage(16..) and highway milage(18..0)
5) The average milage is high for Datsun
6) Petrol is efficently is used for Hyundai and disel is used for tata
7) The correlation between city milage and highway milage is postive
8) Average Seating Capacity: 5.27007874015748 , Number of models with Central Locking: 112 , Number of models with Child Safety Locks: 1201
9) Number of models with High-Speed Alerts: 208 (16.30%), Number of models with Seat Belt Reminders: 1086 (85.11%), Number of models with Door-            Ajar Warnings: 1133 (88.79%)

##### Final Business Conclusion
1). Increased Efficiency: The project improved operational efficiency by automating manual tasks such as inventory management, order processing, and data analysis.     
2). Enhanced Decision-Making: Data-driven insights from analysis, such as trends in sales, employee performance, and customer preferences, helped in making more informed business decisions.